sno/conflicts.py

```diff
@@ -0,0 +1,244 @@
+import sys
+from collections import namedtuple
+
+import click
+
+from .diff_output import repr_row
+from .exceptions import InvalidOperation, NotYetImplemented
+from .structure import RepositoryStructure
+
+
+CommitWithReference = namedtuple("CommitWithReference", ("commit", "reference"))
+
+
+def first_true(iterable):
+    """Returns the value from the iterable that is truthy."""
+    return next(filter(None, iterable))
+
+
+def is_interactive_terminal():
+    if sys.stdin.isatty() and sys.stdout.isatty():
+        return True
+    elif sys.stdin.isatty() and not sys.stdout.isatty():
+        raise InvalidOperation(
+            "Redirecting stdout but not stdin breaks the interactive prompts"
+        )
+    return False
+
+
+def interactive_pause(prompt):
+    """Like click.pause() but waits for the Enter key specifically."""
+    click.prompt(prompt, prompt_suffix="", default="", show_default=False)
+
+
+def is_empty_stream(stream):
+    if stream.seekable():
+        pos = stream.tell()
+        if stream.read(1) == "":
+            return True
+        stream.seek(pos)
+    return False
+
+
+def _get_long_label(commit, reference):
+    """Given a commit and a reference (can be None), returns a label"""
+    if reference is not None:
+        return f'"{reference.shorthand}" ({commit.id.hex})'
+    else:
+        return f"({commit.id.hex})"
+
+
+def _safe_get_dataset_for_index_entry(repo_structure, index_entry):
+    """Gets the dataset that a pygit2.IndexEntry refers to, or None"""
+    try:
+        return repo_structure.get_for_index_entry(index_entry)
+    except KeyError:
+        return None
+
+
+def _get_pk_for_index_entry(dataset, index_entry):
+    """Uses a dataset to decode the primary key that a pygit2.IndexEntry refers to"""
+    return dataset.index_entry_to_pk(index_entry)
+
+
+def _safe_get_feature(dataset, pk):
+    """Gets the dataset's feature with a particular primary key, or None"""
+    try:
+        _, feature = dataset.get_feature(pk)
+        return feature
+    except KeyError:
+        return None
+
+
+# We have three versions of lots of objects - the 3 versions are ancestor, ours, theirs.
+# This namedtuple helps us keep track of them all.
+AncestorOursTheirs = namedtuple("AncestorOursTheirs", ("ancestor", "ours", "theirs"))
+
+AncestorOursTheirs.names = AncestorOursTheirs._fields
+AncestorOursTheirs.chars = tuple(n[0] for n in AncestorOursTheirs.names)
+
+
+def aot(*args):
```

Comment:
fwiw, you can also use namedtuple as a parent class more easily under Py3 by inheriting from `typing.NamedTuple` and then add/override methods/whatever like a normal class.

(what you have is fine though)